tests/functional/src/main/java/org/mule/functional/junit4/ExtensionsTestInfrastructureDiscoverer.java

```diff
@@ -0,0 +1,222 @@
+import org.mule.runtime.extension.api.introspection.ExtensionFactory;
+import org.mule.runtime.extension.api.introspection.declaration.DescribingContext;
+import org.mule.runtime.extension.api.introspection.declaration.spi.Describer;
+import org.mule.runtime.extension.api.resources.GeneratedResource;
+import org.mule.runtime.extension.api.resources.ResourcesGenerator;
+import org.mule.runtime.extension.api.resources.spi.GeneratedResourceFactory;
+import org.mule.runtime.module.extension.internal.DefaultDescribingContext;
+import org.mule.runtime.module.extension.internal.introspection.DefaultExtensionFactory;
+import org.mule.runtime.module.extension.internal.introspection.describer.AnnotationsBasedDescriber;
+import org.mule.runtime.module.extension.internal.introspection.version.StaticVersionResolver;
+import org.mule.runtime.module.extension.internal.manager.DefaultExtensionManager;
+import org.mule.runtime.module.extension.internal.manager.ExtensionManagerAdapter;
+import org.mule.runtime.module.extension.internal.resources.AbstractResourcesGenerator;
+
+import com.google.common.collect.ImmutableList;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.jar.Manifest;
+
+import org.apache.commons.io.FileUtils;
+
+/**
+ * It creates a {@link ExtensionManager} and automatically registers extensions.
+ * <p/>
+ * Once extensions are registered, a {@link ResourcesGenerator} is used to automatically
+ * generate any backing resources needed (XSD schemas, spring bundles, etc).
+ * <p/>
+ * In this way, the user experience is greatly simplified when running the test
+ * either through an IDE or build tool such as maven or gradle.
+ * <p/>
+ *
+ * @since 4.0
+ */
+public class ExtensionsTestInfrastructureDiscoverer
+{
+    private final ServiceRegistry serviceRegistry = new SpiServiceRegistry();
+    private final ExtensionFactory extensionFactory = new DefaultExtensionFactory(serviceRegistry, getClass().getClassLoader());
+    private final ExtensionManagerAdapter extensionManager;
+    private final File generatedResourcesDirectory;
+
+    public ExtensionsTestInfrastructureDiscoverer(ExtensionManagerAdapter extensionManager, File generatedResourcesDirectory)
+    {
+        try
+        {
+            this.extensionManager = extensionManager;
+            this.generatedResourcesDirectory = generatedResourcesDirectory;
+            createManifestFileIfNecessary(generatedResourcesDirectory);
+        }
+        catch (Exception e)
+        {
+            throw new RuntimeException("Error creating discoverer", e);
+        }
+    }
+
+    public ExtensionsTestInfrastructureDiscoverer(File generatedResourcesDirectory)
+    {
+        this(new DefaultExtensionManager(), generatedResourcesDirectory);
+        ((DefaultExtensionManager) extensionManager).setMuleContext(new DefaultMuleContext()
```

Comment:
Done, refactored the code to receive an extension manager instead of creating one.